shared/src/types/key/secp256k1.rs

```diff
@@ -0,0 +1,488 @@
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let vec = hex::decode(s).map_err(ParsePublicKeyError::InvalidHex)?;
+        BorshDeserialize::try_from_slice(&vec)
+            .map_err(ParsePublicKeyError::InvalidEncoding)
+    }
+}
+
+impl From<libsecp256k1::PublicKey> for PublicKey {
+    fn from(pk: libsecp256k1::PublicKey) -> Self {
+        Self(pk)
+    }
+}
+
+/// Secp256k1 secret key
+#[derive(Debug, Clone)]
+pub struct SecretKey(pub libsecp256k1::SecretKey);
+
+impl super::SecretKey for SecretKey {
+    type PublicKey = PublicKey;
+
+    const TYPE: SchemeType = SigScheme::TYPE;
+
+    fn try_from_sk<PK: super::SecretKey>(
+        pk: &PK,
+    ) -> Result<Self, ParseSecretKeyError> {
+        if PK::TYPE == super::common::SecretKey::TYPE {
+            super::common::SecretKey::try_from_sk(pk).and_then(|x| match x {
+                super::common::SecretKey::Secp256k1(epk) => Ok(epk),
+                _ => Err(ParseSecretKeyError::MismatchedScheme),
+            })
+        } else if PK::TYPE == Self::TYPE {
+            Self::try_from_slice(pk.try_to_vec().unwrap().as_slice())
+                .map_err(ParseSecretKeyError::InvalidEncoding)
+        } else {
+            Err(ParseSecretKeyError::MismatchedScheme)
+        }
+    }
+}
+
+impl Serialize for SecretKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> 
+    where
+        S: Serializer,
+    {
+        // not sure if this is how I should be doing this!
+        // https://serde.rs/impl-serialize.html!
+        let arr = self.0.serialize();
+        let mut seq = serializer.serialize_tuple(arr.len())?;
+        for elem in &arr[..] {
+            seq.serialize_element(elem)?;
+        }
+        seq.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for SecretKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>
+    {
+        struct ByteArrayVisitor;
+
+        impl<'de> Visitor<'de> for ByteArrayVisitor {
+            type Value = [u8; libsecp256k1::util::SECRET_KEY_SIZE];
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str(&format!("expecting an array of length {}", libsecp256k1::util::SECRET_KEY_SIZE))
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+                where
+                    A: SeqAccess<'de>,
+            {
+                let mut arr = [0u8; libsecp256k1::util::SECRET_KEY_SIZE];
+                #[allow(clippy::needless_range_loop)]
+                for i in 0..libsecp256k1::util::SECRET_KEY_SIZE {
+                    arr[i] = seq.next_element()?
+                        .ok_or_else(|| Error::invalid_length(i, &self))?;
+                }
+                Ok(arr)
+            }
+        }
+
+        let arr_res = deserializer.deserialize_tuple(libsecp256k1::util::SECRET_KEY_SIZE, ByteArrayVisitor)?;
```

Comment:
Then you can deserialize this back from `[u8; 32]` without the visitor with e.g.:
```suggestion
        let arr_res: [u8; libsecp256k1::util::SECRET_KEY_SIZE] = serde::Deserialize::deserialize(deserializer)?;
```

Reply:
similar for the `Signature` below